### PR TITLE
[Core/Bugfix] Add FP8 K/V Scale and dtype conversion for prefix/prefill Triton Kernel

### LINKED ISSUE #4381: [Bug]: Chunked prefill doesn't seem to work when --kv-cache-dtype fp8
### Your current environment

H100 (but I believe it happens in any machine)

### 🐛 Describe the bug

```
--enable-chunked-prefill --num-max-batched-tokens 2048 --kv-cache-dtype "fp8"
```

Seems to be broken with some type incompatibility error.

### PR DESCRIPTION
**Pull request by jon-chuang**
Fix the FP8 Triton kernel issue. Should enable FP8 KV Cache to be used with:
1. chunked prefill
2. prefix caching

FIX https://github.com/vllm-project/vllm/issues/4381 https://github.com/vllm-project/vllm/issues/3880 https://github.com/vllm-project/vllm/issues/3156 https://github.com/vllm-project/vllm/issues/3880

TODO:
- [x] Undo guards: https://github.com/vllm-project/vllm/pull/3903
- [x] Regression test: https://github.com/vllm-project/vllm/issues/4381#issuecomment-2271579989

Notes:
1. @comaniac mentions upcoming flashinfer support, but I think supporting it in Triton as fallback is good for users who want to avoid installing heavyweight dependency
2. Regarding correctness, you can see that this PR merely brings up triton kernel to parity with vLLM's custom torch CUDA paged attention kernels: https://github.com/vllm-project/vllm/blob/a3bbbfa1d8c2f30581d37c6f30429d648bbbf87c/csrc/attention/attention_kernels.cu#L287,  https://github.com/vllm-project/vllm/blob/a3bbbfa1d8c2f30581d37c6f30429d648bbbf87c/csrc/attention/attention_kernels.cu#L417, where scaled_convert is defined as:
https://github.com/vllm-project/vllm/blob/a3bbbfa1d8c2f30581d37c6f30429d648bbbf87c/csrc/quantization/fp8/nvidia/quant_utils.cuh#L299
---

Example Output (decode v.s. chunked prefill with FP8 KV Cache):
image

Performance is similar to decode in low load case on max_chunk=16,max_sequence_len=512
image
image

I attribute the slightly lower perf to lack of perf tuning of the triton kernels

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

CC: @comaniac

### CODE REVIEW

File: vllm/config.py
Comment (jon-chuang) [inline vllm/config.py:500]: @robertgshaw2-neuralmagic is this sufficient to undo https://github.com/vllm-project/vllm/pull/3903 or am I missing something here?

**Comment by jon-chuang** [comment]
@comaniac how can I trigger the CI? I have no dev env for vllm currently

**Comment by comaniac** [comment]
The CI is already triggered.

**Comment by jon-chuang** [comment]
I guess it is waiting for a runner to schedule `buildkite/ci-aws/pr`?

**Comment by jon-chuang** [comment]
Turns out I actually needed to push new commit after converting from draft status to trigger CI

File: vllm/attention/ops/prefix_prefill.py
Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py]: Not verified, I have no dev env setup unfortunately. Hopefully will set it up soon (today/tmr). For now will try and see.

Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py]: Actually, I think if the KV cache dtype for vLLM is in `uint8`. This code above is definitely wrong if this is the case.

This means we need to do `Tensor.view(dtype=fp8_dtype)` before passing into triton

Comment (comaniac) [reply vllm/attention/ops/prefix_prefill.py]: Right you need to explicitly use `view` to correct the type.

For now I'll remove the ready label to save CI resources. We should add the label back after the PR is verified locally and got reviewed.

Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py]: Ok, I will work on getting my local env set up properly. I used to have a laptop with NVIDIA gpu but now use macbook 😅

**Comment by jon-chuang** [comment]
Hmm, actually; shouldn't Triton have support for mixed FP8 matmul?

https://github.com/triton-lang/triton/blob/194a00f0f54ecd85dba202d840242c5f3f72b068/test/TritonGPU/prefetch.mlir#L14

Anw, this is a story for next time; as we first need to ask users if they want to enable lower precision matmul with one operand in FP8

File: vllm/attention/ops/prefix_prefill.py
Comment (jon-chuang) [inline vllm/attention/ops/prefix_prefill.py:179]: to v.dtype, ditto for k.dtype

File: vllm/config.py
Comment (jon-chuang) [reply vllm/config.py:532]: Actually, upon further testing, prefix caching does not use triton prefix attention code path

**Comment by jon-chuang** [comment]
Should be working now:
image

**Comment by jon-chuang** [comment]
Performance is similar to decode in low load case on max_chunk=16,max_sequence_len=512
image
image

**Comment by jon-chuang** [comment]
@comaniac all tests pass

**Comment by chenxu2048** [comment]
Hi, @jon-chuang. Thanks for your contributions.

The code in #3234 is too old to merge. I'll close #3234 once this PR done.

File: vllm/attention/ops/prefix_prefill.py
Comment (chenxu2048) [inline vllm/attention/ops/prefix_prefill.py:741]: Maybe we could let k_cache and v_cache in real type, and remove the conversion and checking.

Here are the discussions about kv cache dtype in https://github.com/vllm-project/vllm/pull/3234#discussion_r1522561127.

Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py:741]: No. Other kernels depend on uint8

Comment (chenxu2048) [reply vllm/attention/ops/prefix_prefill.py:741]: Which kernel?

Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py:741]: Other code may depend implicitly on this behaviour. This refactor to store fp8 KV cache not in uint8 dtype is unrelated to this PR and prefer if it is done separately.

Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py:741]: Created a PR here: https://github.com/vllm-project/vllm/pull/7302

File: tests/kernels/test_prefix_prefill.py
Comment (jon-chuang) [inline tests/kernels/test_prefix_prefill.py:60]: I think you should parametrize by "fp8" dtypes here

Comment (jon-chuang) [reply tests/kernels/test_prefix_prefill.py]: Done, additional addition to testing path

**Comment by JaheimLee** [comment]
Can not work on NVIDIA Ampere GPU, like 3090.

**Comment by jon-chuang** [comment]
> Can not work on NVIDIA Ampere GPU, for example 3090.

Unfortunate limit of Triton

**Comment by JaheimLee** [comment]
> > Can not work on NVIDIA Ampere GPU, for example 3090.
> 
> Unfortunate limit of Triton

Does [#5975](https://github.com/vllm-project/vllm/pull/5975) help for this?

**Comment by comaniac** [comment]
> > > Can not work on NVIDIA Ampere GPU, for example 3090.
> > 
> > 
> > Unfortunate limit of Triton
> 
> Does [#5975](https://github.com/vllm-project/vllm/pull/5975) help for this?

No it's for GEMM, not attention.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:20]: We should not allow known potential illegal memory access issue. At least we need to know what workloads/models would likely trigger that, and have an issue for people to report and track progress.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:75]: I feel you could just add this parameter to the above test so that you don't have to duplicate an entire test function.

File: tests/kernels/test_prefix_prefill.py
Comment (comaniac) [inline tests/kernels/test_prefix_prefill.py:224]: nit
```suggestion
    atol = 1e-3 if "fp8" in kv_cache_dtype else 1e-6
    torch.testing.assert_close(output, output_ref, atol=atol, rtol=0)
```

File: vllm/attention/ops/prefix_prefill.py
Comment (comaniac) [inline vllm/attention/ops/prefix_prefill.py:740]: Blocked by #7302

**Review by comaniac — COMMENTED** [review]
Otherwise LGTM after merging #8302 and passing the CI

**Comment by comaniac** [comment]
btw did you test on H100?

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:75]: I tried this but as you notice, I'm not using hf to validate but a completely different way (validate against the non chunked prefill version and only the log probs, not the generation)

Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:75]: The reason is because of the Numerical error due to fp8 quant

**Comment by jon-chuang** [comment]
Yes tested locally on H100.

File: vllm/attention/ops/prefix_prefill.py
Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py:740]: Yes we can land this first and get rid of it in #7302 depending on merge order

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py:75]: You can still merge them to one test and just use different checks for different types?
(Also please comment on the test for reasons.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:20]: Should I file an issue for it?

Well landing this is still better than error for every model, but you're right we should track it.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:75]: Sure but it's kind of swiss cheesing at that point as all of the arguments and output variables are different

If you want I can name this test `test_models_log_probs` `while the other is test_models_exact_match`

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py:20]: Yes please file an issue. We don't want users to randomly see this error but cannot find any related resource.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py:75]: Ok how about we keep this test but
1. rename to something like `test_fp8_model_and_kv_cache`, and
2. add docstring to explain why we compare logprobs.

Also for `facebook/opt-125m`, since you're using FP16 checkpoint which doesn't have kv cache scaling factors, you should test e5m2 for better accuracy.

Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:75]: Ok sounds good, maybe that was the reason for the illegal memory access issue too let me check.

Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: Unfortunately; still illegal memory access

Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: Unfortunately, still illegal memory access

File: vllm/attention/ops/prefix_prefill.py
Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py:740]: @comaniac I think #7302 is more controversial; prefer to merge this PR first if everything checks out. WDYT?

Follow up upon merge is post issue about the illegal memory access

Specifically one of these errors:

```python
RuntimeError: Triton Error [CUDA]: an illegal memory access was encountered
RuntimeError: CUDA error: an illegal memory access was encountered
```

Comment (jon-chuang) [reply vllm/attention/ops/prefix_prefill.py:740]: I closed #7302 by virtue of being unnecessary complexity at no benefit. This last-mile dispatch is fine IMO

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: Created issue: https://github.com/vllm-project/vllm/issues/7378

**Comment by jon-chuang** [comment]
@comaniac should be ready for merge

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:13]: Not necessary to call them exact match models.
```suggestion
MODELS = [
```

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:22]: Does other models work with E5M2? For example does that work if you just use the FP16 `Qwen2-1.5B-Instruct` checkpoint?

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:38]: I now feel that we don't have to rename this test. Just leave it here to indicate that we are testing the most common use case (i.e., fp16).

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:112]: I would suggest the following organization:
1. The model list includes `MODELS` and `FP8_KV_MODELS`.
2. For `fp8_e5m2` kv-cache, simply test with `MODELS`.
3. For `fp8_e4m3` kv-cache, test with `FP8_KV_MODELS`.

So that you could have

```python
@pytest.mark.parametrize("model_n_kv_dtype", [("fp8_e5m2", MODELS), ("fp8_e4m3", FP8_KV_MODELS)])
...
def test_model_with_fp8_kv_cache(model_n_kv_dtype, ...):
    kv_cache_dtype, model = model_n_kv_dtype
    ....
```

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:120]: This will never be -1?

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:117]: Seems not necessary as chunked_prefill_token_size is at most 16.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:149]:
```suggestion
        name_0="no_chunked_prefill",
```

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:115]: Why you need to compute 32 max new tokens but only check the first 4? You should just configure the max_tokens to be the number of tokens you want to check.

File: vllm/attention/ops/prefix_prefill.py
Comment (comaniac) [inline vllm/attention/ops/prefix_prefill.py:22]:
```suggestion
        k_scale,
        v_scale,
```

File: vllm/attention/ops/prefix_prefill.py
Comment (comaniac) [inline vllm/attention/ops/prefix_prefill.py:456]:
```suggestion
        k_scale,
        v_scale,
```

File: vllm/attention/ops/prefix_prefill.py
Comment (comaniac) [inline vllm/attention/ops/prefix_prefill.py:727]:
```suggestion
            if kv_cache_dtype in ("fp8", "fp8_e4m3"):
```

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:120]: supposed to be 1.. not sure what happened

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: I think models for which the KV has not been explicitly quantized there can be significant numerical error, even though the outputs are sensible. Hence I think it is better to only test models that have been quantized

If you want I can include it to exercise the code path but it is not possible to check correctness IMO.

Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: I will also add quantized llama-2 KV fp8 to testing

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: I can't seem to find any model that has quantized kv E5M2; if not explicitly quantized with KV scales, the log probs differ due to the inaccuracy during low-precision float conversion

Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py]: E5M2 doesn't need scaling factors and the accuracy shouldn't drop significantly. If it doesn't make sense to check the logprobs for E5M2, we could skip testing them.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py]: The only reason I proposed this is because you added opt-125m which actually tests E5M2 without scaling factors.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: I see, the reason the test failed was because Qwen2-1.5B is actually BF16.
I tried with llama-2 which is FP16 and the tests pass now/

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py]: See above; now it is tested with llama-2-7b e5m2

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:30]: Feel no need to introduce this. Just specify the number in `max_tokens` and check them all.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [inline tests/basic_correctness/test_chunked_prefill.py:87]: Why not just using a tuple?

**Review by comaniac — COMMENTED** [review]
Last batch of minor comments

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:124]: I'm not sure I was following the previous test blindly, which is totally wrong as you mentioned. If you want I can get rid of the 1 for both.

Alternately I can also remove the if statement which will allow testing the code path with `enable_chunked_prefill=True, prefill_size=1`

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:30]: ok

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:87]: Because in the test outputs it shows `kv_dtype_n_model0` which is totally not readable and diagnosable. This will show the full string.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:124]: I remove the if statement from both which is useless as you said

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py:82]: I know this issue, but I don't think it's worthwhile to use this hacky way. You should be able to easily find out what `kv_dtype_n_model0` means by looking at the test, which you have to do it anyways when you're debugging.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (comaniac) [reply tests/basic_correctness/test_chunked_prefill.py]: Yeah let's just make it reasonable. Also cc @rkooo567 who may know what it forms like this.

File: tests/basic_correctness/test_chunked_prefill.py
Comment (jon-chuang) [reply tests/basic_correctness/test_chunked_prefill.py:82]: Actually, i realize the way to do it, you can just write the key to parameterize as comma separated, and it will parse not as tuple but individual. Thanks

**Review by comaniac — APPROVED** [review]
LGTM

**Comment by jon-chuang** [comment]
Thank you @comaniac. Need your help to merge as I have no write access.

**Comment by pavanimajety** [comment]
@jon-chuang Is this to-do fixed with this MR? 

https://github.com/vllm-project/vllm/blame/main/vllm/attention/backends/xformers.py#L601

**Comment by jon-chuang** [comment]
Yes, the comment should have been removed

**Comment by jon-chuang** [comment]
Yes @chenxu2048

**Pull request merged**
